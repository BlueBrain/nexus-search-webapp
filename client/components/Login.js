import React from "react";
import { WithStore } from "@bbp/nexus-react";
import { auth } from "../store/actions";
import { Menu, Dropdown, Icon } from "antd";
import CopyToClipboard from "./CopyToClipboard";
import { Link } from "react-router-dom";

const userMenu = (logout, token) => (
  <Menu>
    <Menu.Item key="0" disabled={true}>
      <CopyToClipboard text={"copy token to clipboard"} value={token}>
        <a target="_blank" rel="noopener noreferrer">
          <Icon type="copy" /> copy token{" "}
        </a>
      </CopyToClipboard>
    </Menu.Item>
    <Menu.Divider />
    <Menu.Item key="3">
      <a href="#" onClick={logout}>
        <Icon type="logout" /> logout
      </a>
    </Menu.Item>
  </Menu>
);

const Login = () => (
  <div className="login-block">
    <WithStore
      mapStateToProps={({ auth, config }) => ({
        name: auth.tokenOwner,
        token: auth.token,
        loginURI: config.loginURI
      })}
      mapDispatchToProps={{
        authenticate: auth.authenticate,
        logout: () => {
          auth.logout();
          location.reload();
        }
      }}
    >
      {({ name, loginURI, logout, token }) => {
        if (token) {
          return (
            <Dropdown overlay={userMenu(logout, token)}>
              <a className="ant-dropdown-link align-center">
                {" "}
                {name} <Icon type="down" />
              </a>
            </Dropdown>
          );
        }
        return (
          <a className="ant-dropdown-link align-center" href={loginURI}>
            Log in <Icon type="login" />
          </a>
        );
      }}
    </WithStore>
  </div>
);

export default Login;
