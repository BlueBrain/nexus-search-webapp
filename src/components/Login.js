import React from "react";
import { Auth, WithStore } from "@bbp/nexus-react";
import { auth } from "../store/actions";

const Login = () => (
  <div className="login-block">
    <WithStore
      mapStateToProps={({ auth, config }) => ({
        name: auth.tokenOwner,
        loginURI: config.loginURI
      })}
      mapDispatchToProps={{
        authenticate: auth.authenticate,
        logout: auth.logout
      }}
    >
      {({ name, loginURI, authenticate, logout }) => (
        <Auth.Container
          name={name}
          authenticate={() => authenticate(window.location.href)}
          logout={logout}
          loginURI={loginURI}
        />
      )}
    </WithStore>
  </div>
);

export default Login;
