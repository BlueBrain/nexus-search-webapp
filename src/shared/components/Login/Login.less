@import '../../lib.less';

.Login {
  background-color: #fff;
  margin: auto;
  width: 300px;
  margin-top: 10%;

  .logo {
    width: 100px;
    margin: auto;
    padding: 2em 1em;
  }

  .message {
    text-align: center;
  }

  .realm-holder {
    padding-top: 0.5em;
  }

  .actions {
    display: block;
    text-align: center;
    padding-top: 0.3rem;
  }

  .realm {
    color: @primary-color;
    text-decoration: underline;
    text-decoration-style: dotted;
    cursor: pointer;
  }
}
