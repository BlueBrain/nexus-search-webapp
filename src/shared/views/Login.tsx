import * as React from 'react';
import { connect } from 'react-redux';
import LoginBox, { Realm } from '../components/Login';
import { AuthState } from '../store/reducers/auth';
import { StaticRouterProps, Redirect } from 'react-router';
import { push } from 'connected-react-router';

export interface LoginViewProps {
  authorizationEndpoint: string;
  realms: Realm[];
  clientId: string;
  hostName: string;
  redirectUrl: string;
  redirect(): void;
}

const Login: React.FunctionComponent<LoginViewProps> = props => {
  const { realms, redirect } = props;
  if (realms.length === 0 || !realms) {
    redirect();
    return null;
  }
  return (
    <LoginBox
      realms={realms}
      clientId={props.clientId}
      hostName={props.hostName}
      redirectUrl={props.redirectUrl}
    />
  );
};

const mapStateToProps = ({
  auth,
  router,
}: {
  auth: AuthState;
  router: StaticRouterProps;
}) => {
  const location = router.location as { state?: { previousUrl: string } };

  return {
    authorizationEndpoint: auth.authorizationEndpoint || '',
    realms: (auth.realms && auth.realms.data && auth.realms.data.results) || [],
    redirectUrl:
      (location && location.state && location.state.previousUrl) ||
      auth.redirectHostName ||
      '',
    hostName: auth.redirectHostName || '',
    clientId: auth.clientId || '',
  };
};

const mapDispatchToProps = (dispatch: any) => ({
  redirect: () => dispatch(push('/')),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Login);
