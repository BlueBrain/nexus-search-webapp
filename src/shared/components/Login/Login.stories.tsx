import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';

import Login from './index';

const realms: string[] = ['BBP', 'HBP', 'Google'];

storiesOf('Components/Login', module).add(
  'Login',
  withInfo(`
    The Login component allows to redirect to the Auth Provider page.

    ~~~js
      <Login loginURL="https://keycloak.org?redirect=https://bbp.epfl.ch/nexus/" />
    ~~~
  `)(() => (
    <React.Fragment>
      <div style={{ margin: '50px 40px 0px' }}>
        <h2>1 Realm to choose from:</h2>
        <Login realms={[realms[0]]} />
      </div>
      <div style={{ margin: '50px 40px 0px' }}>
        <h2>3 Realms to choose from:</h2>
        <Login realms={realms} />
      </div>
    </React.Fragment>
  ))
);
