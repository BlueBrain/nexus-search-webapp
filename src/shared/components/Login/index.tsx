import * as React from 'react';
import { Card, Dropdown, Icon, Menu, Button } from 'antd';

import './Login.less';

const logo = require('../../logo.svg');

export interface LoginProps {
  realms: string[];
  selectedRealm?: string;
  onLogin?(e: React.SyntheticEvent): void;
  onRealmSelected?(name: string): void;
}

const Login: React.FunctionComponent<LoginProps> = ({
  realms,
  selectedRealm,
  onLogin = () => {},
  onRealmSelected = () => {},
}) => {
  const [realm, setRealm] = React.useState(selectedRealm || realms[0]);

  const menu = (
    <Menu
      onClick={({ key, domEvent }) => {
        domEvent.preventDefault();
        domEvent.stopPropagation();
        const realm = realms.find(r => r === key);
        if (realm) {
          setRealm(realm);
          onRealmSelected(realm);
        }
      }}
    >
      {realms.map(realm => (
        <Menu.Item key={realm}>{realm}</Menu.Item>
      ))}
    </Menu>
  );

  return (
    <div className="Login">
      <Card
        cover={<img className="logo" alt="Nexus logo" src={logo} />}
        size="small"
        bodyStyle={{ borderTop: '1px solid rgba(0, 0, 0, 0.10)' }}
      >
        {realms.length === 1 ? (
          <Button
            className="login-button"
            block
            onClick={onLogin}
            type="primary"
          >
            Log in
            <Icon type="login" />
          </Button>
        ) : (
          <div className="actions">
            <Button className="login-button" onClick={onLogin} type="primary">
              Log in
              <Icon type="login" />
            </Button>
            <div className="realm-holder">
              <span> with </span>
              <Dropdown overlay={menu} trigger={['click', 'hover']}>
                <span className="realm">
                  {realm} <Icon type="down" />
                </span>
              </Dropdown>
            </div>
          </div>
        )}
      </Card>
    </div>
  );
};

export default Login;
