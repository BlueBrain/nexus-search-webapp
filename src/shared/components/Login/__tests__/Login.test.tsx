import * as React from 'react';
import { mount } from 'enzyme';

import Login from '../index';

const realms: string[] = ['BBP', 'HBP', 'Google'];
const loginComponent = <Login realms={[realms[1]]} />;
const wrapper = mount(loginComponent);

describe('login component', () => {
  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('with only 1 realm', () => {
    it('should have a login button', () => {
      expect(wrapper.find('button.login-button').text()).toEqual('Log in');
    });
  });

  describe('with more than 1 realm', () => {
    beforeAll(() => {
      // change props and pass all realms
      wrapper.setProps({ realms });
    });

    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should have a login button', () => {
      expect(wrapper.find('button.login-button').text()).toEqual('Log in');
    });

    it("a dropdown should show a Realm's name", () => {
      // There is an extra space appended to because of the dropdown Icon
      expect(wrapper.find('span.realm').text()).toEqual('HBP ');
    });
  });
});
