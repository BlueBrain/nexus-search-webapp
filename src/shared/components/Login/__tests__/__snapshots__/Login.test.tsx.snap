// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login component should render correctly 1`] = `
<Login
  realms={
    Array [
      "HBP",
    ]
  }
>
  <div
    className="Login"
  >
    <Card
      bodyStyle={
        Object {
          "borderTop": "1px solid rgba(0, 0, 0, 0.10)",
        }
      }
      cover={
        <img
          alt="Nexus logo"
          className="logo"
          src="/mock/path"
        />
      }
      size="small"
    >
      <div
        className="ant-card ant-card-bordered ant-card-small"
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="Nexus logo"
            className="logo"
            src="/mock/path"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "borderTop": "1px solid rgba(0, 0, 0, 0.10)",
            }
          }
        >
          <Button
            block={true}
            className="login-button"
            ghost={false}
            htmlType="button"
            loading={false}
            onClick={[Function]}
            type="primary"
          >
            <Wave>
              <button
                className="ant-btn login-button ant-btn-primary ant-btn-block"
                onClick={[Function]}
                type="button"
              >
                <span
                  key=".0"
                >
                  Log in
                </span>
                <Icon
                  key=".1"
                  type="login"
                >
                  <LocaleReceiver
                    componentName="Icon"
                  >
                    <i
                      aria-label="icon: login"
                      className="anticon anticon-login"
                    >
                      <IconReact
                        className=""
                        type="login-o"
                      >
                        <svg
                          aria-hidden="true"
                          className=""
                          data-icon="login"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          key="svg-login"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M521.7 82c-152.5-.4-286.7 78.5-363.4 197.7-3.4 5.3.4 12.3 6.7 12.3h70.3c4.8 0 9.3-2.1 12.3-5.8 7-8.5 14.5-16.7 22.4-24.5 32.6-32.5 70.5-58.1 112.7-75.9 43.6-18.4 90-27.8 137.9-27.8 47.9 0 94.3 9.3 137.9 27.8 42.2 17.8 80.1 43.4 112.7 75.9 32.6 32.5 58.1 70.4 76 112.5C865.7 417.8 875 464.1 875 512c0 47.9-9.4 94.2-27.8 137.8-17.8 42.1-43.4 80-76 112.5s-70.5 58.1-112.7 75.9A352.8 352.8 0 0 1 520.6 866c-47.9 0-94.3-9.4-137.9-27.8A353.84 353.84 0 0 1 270 762.3c-7.9-7.9-15.3-16.1-22.4-24.5-3-3.7-7.6-5.8-12.3-5.8H165c-6.3 0-10.2 7-6.7 12.3C234.9 863.2 368.5 942 520.6 942c236.2 0 428-190.1 430.4-425.6C953.4 277.1 761.3 82.6 521.7 82zM395.02 624v-76h-314c-4.4 0-8-3.6-8-8v-56c0-4.4 3.6-8 8-8h314v-76c0-6.7 7.8-10.5 13-6.3l141.9 112a8 8 0 0 1 0 12.6l-141.9 112c-5.2 4.1-13 .4-13-6.3z"
                            key="svg-login-svg-0"
                          />
                        </svg>
                      </IconReact>
                    </i>
                  </LocaleReceiver>
                </Icon>
              </button>
            </Wave>
          </Button>
        </div>
      </div>
    </Card>
  </div>
</Login>
`;

exports[`login component with more than 1 realm should render correctly 1`] = `
<Login
  realms={
    Array [
      "BBP",
      "HBP",
      "Google",
    ]
  }
>
  <div
    className="Login"
  >
    <Card
      bodyStyle={
        Object {
          "borderTop": "1px solid rgba(0, 0, 0, 0.10)",
        }
      }
      cover={
        <img
          alt="Nexus logo"
          className="logo"
          src="/mock/path"
        />
      }
      size="small"
    >
      <div
        className="ant-card ant-card-bordered ant-card-small"
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="Nexus logo"
            className="logo"
            src="/mock/path"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "borderTop": "1px solid rgba(0, 0, 0, 0.10)",
            }
          }
        >
          <div
            className="actions"
          >
            <Button
              block={false}
              className="login-button"
              ghost={false}
              htmlType="button"
              loading={false}
              onClick={[Function]}
              type="primary"
            >
              <Wave>
                <button
                  className="ant-btn login-button ant-btn-primary"
                  onClick={[Function]}
                  type="button"
                >
                  <span
                    key=".0"
                  >
                    Log in
                  </span>
                  <Icon
                    key=".1"
                    type="login"
                  >
                    <LocaleReceiver
                      componentName="Icon"
                    >
                      <i
                        aria-label="icon: login"
                        className="anticon anticon-login"
                      >
                        <IconReact
                          className=""
                          type="login-o"
                        >
                          <svg
                            aria-hidden="true"
                            className=""
                            data-icon="login"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            key="svg-login"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M521.7 82c-152.5-.4-286.7 78.5-363.4 197.7-3.4 5.3.4 12.3 6.7 12.3h70.3c4.8 0 9.3-2.1 12.3-5.8 7-8.5 14.5-16.7 22.4-24.5 32.6-32.5 70.5-58.1 112.7-75.9 43.6-18.4 90-27.8 137.9-27.8 47.9 0 94.3 9.3 137.9 27.8 42.2 17.8 80.1 43.4 112.7 75.9 32.6 32.5 58.1 70.4 76 112.5C865.7 417.8 875 464.1 875 512c0 47.9-9.4 94.2-27.8 137.8-17.8 42.1-43.4 80-76 112.5s-70.5 58.1-112.7 75.9A352.8 352.8 0 0 1 520.6 866c-47.9 0-94.3-9.4-137.9-27.8A353.84 353.84 0 0 1 270 762.3c-7.9-7.9-15.3-16.1-22.4-24.5-3-3.7-7.6-5.8-12.3-5.8H165c-6.3 0-10.2 7-6.7 12.3C234.9 863.2 368.5 942 520.6 942c236.2 0 428-190.1 430.4-425.6C953.4 277.1 761.3 82.6 521.7 82zM395.02 624v-76h-314c-4.4 0-8-3.6-8-8v-56c0-4.4 3.6-8 8-8h314v-76c0-6.7 7.8-10.5 13-6.3l141.9 112a8 8 0 0 1 0 12.6l-141.9 112c-5.2 4.1-13 .4-13-6.3z"
                              key="svg-login-svg-0"
                            />
                          </svg>
                        </IconReact>
                      </i>
                    </LocaleReceiver>
                  </Icon>
                </button>
              </Wave>
            </Button>
            <div
              className="realm-holder"
            >
              <span>
                 with 
              </span>
              <Dropdown
                mouseEnterDelay={0.15}
                mouseLeaveDelay={0.1}
                overlay={
                  <Menu
                    onClick={[Function]}
                  >
                    <MenuItem>
                      BBP
                    </MenuItem>
                    <MenuItem>
                      HBP
                    </MenuItem>
                    <MenuItem>
                      Google
                    </MenuItem>
                  </Menu>
                }
                placement="bottomLeft"
                trigger={
                  Array [
                    "click",
                    "hover",
                  ]
                }
              >
                <Dropdown
                  defaultVisible={false}
                  mouseEnterDelay={0.15}
                  mouseLeaveDelay={0.1}
                  onVisibleChange={[Function]}
                  overlay={[Function]}
                  overlayClassName=""
                  overlayStyle={Object {}}
                  placement="bottomLeft"
                  prefixCls="ant-dropdown"
                  showAction={Array []}
                  transitionName="slide-up"
                  trigger={
                    Array [
                      "click",
                      "hover",
                    ]
                  }
                >
                  <Trigger
                    action={
                      Array [
                        "click",
                        "hover",
                      ]
                    }
                    afterPopupVisibleChange={[Function]}
                    blurDelay={0.15}
                    builtinPlacements={
                      Object {
                        "bottomCenter": Object {
                          "offset": Array [
                            0,
                            4,
                          ],
                          "overflow": Object {
                            "adjustX": 1,
                            "adjustY": 1,
                          },
                          "points": Array [
                            "tc",
                            "bc",
                          ],
                          "targetOffset": Array [
                            0,
                            0,
                          ],
                        },
                        "bottomLeft": Object {
                          "offset": Array [
                            0,
                            4,
                          ],
                          "overflow": Object {
                            "adjustX": 1,
                            "adjustY": 1,
                          },
                          "points": Array [
                            "tl",
                            "bl",
                          ],
                          "targetOffset": Array [
                            0,
                            0,
                          ],
                        },
                        "bottomRight": Object {
                          "offset": Array [
                            0,
                            4,
                          ],
                          "overflow": Object {
                            "adjustX": 1,
                            "adjustY": 1,
                          },
                          "points": Array [
                            "tr",
                            "br",
                          ],
                          "targetOffset": Array [
                            0,
                            0,
                          ],
                        },
                        "topCenter": Object {
                          "offset": Array [
                            0,
                            -4,
                          ],
                          "overflow": Object {
                            "adjustX": 1,
                            "adjustY": 1,
                          },
                          "points": Array [
                            "bc",
                            "tc",
                          ],
                          "targetOffset": Array [
                            0,
                            0,
                          ],
                        },
                        "topLeft": Object {
                          "offset": Array [
                            0,
                            -4,
                          ],
                          "overflow": Object {
                            "adjustX": 1,
                            "adjustY": 1,
                          },
                          "points": Array [
                            "bl",
                            "tl",
                          ],
                          "targetOffset": Array [
                            0,
                            0,
                          ],
                        },
                        "topRight": Object {
                          "offset": Array [
                            0,
                            -4,
                          ],
                          "overflow": Object {
                            "adjustX": 1,
                            "adjustY": 1,
                          },
                          "points": Array [
                            "br",
                            "tr",
                          ],
                          "targetOffset": Array [
                            0,
                            0,
                          ],
                        },
                      }
                    }
                    defaultPopupVisible={false}
                    defaultVisible={false}
                    destroyPopupOnHide={false}
                    focusDelay={0}
                    getDocument={[Function]}
                    getPopupClassNameFromAlign={[Function]}
                    hideAction={Array []}
                    mask={false}
                    maskClosable={true}
                    mouseEnterDelay={0.15}
                    mouseLeaveDelay={0.1}
                    onPopupAlign={[Function]}
                    onPopupVisibleChange={[Function]}
                    onVisibleChange={[Function]}
                    overlay={[Function]}
                    popup={[Function]}
                    popupAlign={Object {}}
                    popupClassName=""
                    popupPlacement="bottomLeft"
                    popupStyle={Object {}}
                    popupTransitionName="slide-up"
                    popupVisible={false}
                    prefixCls="ant-dropdown"
                    showAction={Array []}
                  >
                    <span
                      className="realm ant-dropdown-trigger"
                      key="trigger"
                      onClick={[Function]}
                      onMouseDown={[Function]}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      onTouchStart={[Function]}
                    >
                      HBP
                       
                      <Icon
                        type="down"
                      >
                        <LocaleReceiver
                          componentName="Icon"
                        >
                          <i
                            aria-label="icon: down"
                            className="anticon anticon-down"
                          >
                            <IconReact
                              className=""
                              type="down-o"
                            >
                              <svg
                                aria-hidden="true"
                                className=""
                                data-icon="down"
                                fill="currentColor"
                                focusable="false"
                                height="1em"
                                key="svg-down"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                                  key="svg-down-svg-0"
                                />
                              </svg>
                            </IconReact>
                          </i>
                        </LocaleReceiver>
                      </Icon>
                    </span>
                  </Trigger>
                </Dropdown>
              </Dropdown>
            </div>
          </div>
        </div>
      </div>
    </Card>
  </div>
</Login>
`;
